### PR TITLE
Merge WebSocketFlow into HTTPFlow, add WebSocket UI

### LINKED ISSUE #4425: Combine WebSocketFlow and HTTPFlow
HTTPFlow and WebSocketFlow currently have a really ugly relation where one depends on the other, they appear separately in dumps, etc. To fix this madness, I would propose we merge WebSocketFlow and HTTPFlow: HTTPFlow just gets a `.websocket: Optional[WebSocketData]` attribute, which contains everything related to the websocket connection (messages, close reason, etc). There also is prior art for this approach: For example, the Chrome Devtools use exactly that data model.

@mitmproxy/devs, thoughts?

### PR DESCRIPTION
**Pull request by mhils**
This PR fixes #4425 and then adds WebSocket Message display in mitmprox as a cherry on top! 🍒 

image

### CODE REVIEW

File: mitmproxy/addons/dumper.py [hunk 1]
```diff
@@ -277,37 +277,37 @@ def error(self, f):
...
-            "Error in WebSocket connection to {}: {}".format(
-                human.format_address(f.server_conn.address), f.error
-            ),
+            f"Error in WebSocket connection to {human.format_address(f.server_conn.address)}: {f.error}",
             fg="red"
         )
 
-    def websocket_message(self, f):
+    def websocket_message(self, f: http.HTTPFlow):
+        assert f.websocket is not None
```
Comment (Prinzhorn) [inline mitmproxy/addons/dumper.py:287]: Are these assertion necessary? If `websocket_message` is called without a websocket then the add-on core is broken, right?

File: mitmproxy/addons/save.py [hunk 2]
```diff
@@ -88,28 +89,24 @@ def tcp_end(self, flow):
...
 
-    def request(self, flow):
+    def request(self, flow: http.HTTPFlow):
         if self.stream:
             self.active_flows.add(flow)
 
-    def response(self, flow):
-        if self.stream:
+    def response(self, flow: http.HTTPFlow):
+        if self.stream and flow.websocket is None:
```
Comment (Prinzhorn) [inline mitmproxy/addons/save.py:105]: Does this mean the HTTP 101 flow is not saved and lost? But it contains interesting data such as authentication tokens. Or why was that change needed?

File: mitmproxy/websocket.py [hunk 3]
```diff
@@ -1,168 +1,136 @@
...
     from_client: bool
     """True if this messages was sent by the client."""
-    content: Union[bytes, str]
+    is_text: bool
+    """
+    True if the message is a text message, False if the message is a binary message.
+
+    In either case, mitmproxy will store the message contents as *bytes*.
+    """
+    content: bytes
```
Comment (Prinzhorn) [inline mitmproxy/websocket.py:40]: What's the reason we're switching to bytes for text messages as well? We know that WebSocket text is UTF-8, doesn't that make it more awkward to work with? Or is that my lack of Python?

File: mitmproxy/websocket.py [hunk 4]
```diff
@@ -1,168 +1,136 @@
...
 
     def set_state(self, state):
-        self.type, self.from_client, self.content, self.timestamp, self.killed = state
-        self.type = Opcode(self.type)  # replace enum with bare int
+        self.from_client, self.is_text, self.content, self.timestamp = state
 
     def __repr__(self):
-        if self.type == Opcode.TEXT:
-            return "text message: {}".format(repr(self.content))
+        if self.is_text:
```
Comment (Prinzhorn) [inline mitmproxy/websocket.py:68]: Currently in my add-on I'm explicitly checking both opcodes because the docs say:

> Currently there are two types of messages, corresponding to the BINARY and TEXT frame types.

and I didn't want to run in a future scenario were my `else` would catch others in a later version.

Does this change conclude that we'll never expose PING/PONG/CLOSE? I guess that's the only reason the opcode was exposed originally?

File: mitmproxy/websocket.py [hunk 5]
```diff
@@ -1,168 +1,136 @@
...
             "WebSocketMessage.kill is deprecated, set an empty content instead.",
             DeprecationWarning,
             stacklevel=2,
         )
-        # empty str or empty bytes.
-        self.content = type(self.content)()
+        self.content = b""
 
+    @property
+    def killed(self) -> bool:  # pragma: no cover
```
Comment (Prinzhorn) [inline mitmproxy/websocket.py:87]: I only just realized that `kill` is also deprecated. I disagree with that decision, but I don't know what lead to that? Can you link to some discussion?

If I set the message to an empty string the original content is lost. In the UI it's just an empty message? But I want the user to know "Hey this message _would_ have been sent, but wasn't". In the same way we don't empty all HTTP headers and body if a flow is dropped. I don't want to lose any information.

File: mitmproxy/websocket.py (hunk 3 above)
Comment (mhils) [reply mitmproxy/websocket.py:40]: I haven't really made my mind up here to be honest. The previous implementation was annoying me quite a bit because `content` was `Union[str,bytes]`, and you always had to add a special case and check. Now my hope is that it's just more consistent to deal with, and if you need text you can just call `.decode()` yourself. @Kriechi, any opinions from your end?

File: mitmproxy/addons/save.py (hunk 2 above)
Comment (Prinzhorn) [reply mitmproxy/addons/save.py:105]: For a moment I must've forgotten that `WebSocketFlow` is gone, makes sense now

File: mitmproxy/websocket.py (hunk 3 above)
Comment (Prinzhorn) [reply mitmproxy/websocket.py:40]: Déjà vu

https://github.com/mitmproxy/mitmproxy/issues/4131

> I remotely recall something about WebSocket TEXT frames having a very strictly defined encoding? It looks like we just forward what we get from wsproto. 

---

> and you always had to add a special case and check

How is that special case gone? You still need to decide when to `.decode()` and when not to.

I'm not favoring one solution or the other, but given that I've brought this exact topic up before maybe we can finally settle this :wink:

Comment (Prinzhorn) [reply mitmproxy/websocket.py:40]: I swear this is the last comment: maybe my original idea of a `text` property kind of makes sense? It would just be a getter/setter that does the decode/encode for you. Otherwise every user would require the knowledge that WebSocket text frames are always utf-8 to manually do the `decode` of the bytes.

If `not is_text` the `text` property should crash or something tho? Because reading `text` for arbitrary `content` won't always work.

File: mitmproxy/websocket.py (hunk 4 above)
Comment (mhils) [reply mitmproxy/websocket.py:68]: Good point, we may want to leave that door open. I'll change it back to using the opcode. 👍

File: mitmproxy/websocket.py (hunk 5 above)
Comment (mhils) [reply mitmproxy/websocket.py:87]: There is none, but there probably should have been some - I was just ripping out as much complexity as I could find. You mention a valid use case though. Going back to the drawing board, what would you expect `kill` to do? Promptly abort the connection, or just suppress the current message? I just now realize that we also don't have a good story for interception either.

File: mitmproxy/proxy/layers/http/__init__.py [hunk 6]
```diff
@@ -308,6 +309,25 @@ def send_response(self, already_streamed: bool = False):
         """We have either consumed the entire response from the server or the response was set by an addon."""
         assert self.flow.response
         self.flow.response.timestamp_end = time.time()
+
+        is_websocket = (
+            self.flow.response.status_code == 101
+            and
+            self.flow.response.headers.get("upgrade", "").lower() == "websocket"
+            and
+            self.flow.request.headers.get("Sec-WebSocket-Version", "") == "13"
```
Comment (Kriechi) [inline mitmproxy/proxy/layers/http/__init__.py:318]: Should we use wsproto constants here instead?

File: test/mitmproxy/proxy/layers/test_websocket.py [hunk 7]
```diff
@@ -95,12 +94,13 @@ def test_upgrade(tctx):
...
-    assert len(flow().messages) == 2
-    assert flow().messages[0].content == "hello world"
-    assert flow().messages[0].from_client
-    assert flow().messages[1].content == b"hello back"
-    assert flow().messages[1].from_client is False
+    assert len(flow().websocket.messages) == 2
+    assert flow().websocket.messages[0].content == b"hello world"
+    assert flow().websocket.messages[0].from_client
+    assert flow().websocket.messages[0].is_text
+    assert flow().websocket.messages[1].content == b"hello back"
```
[the remark this answers is not in the document]
Comment (Prinzhorn) [reply test/mitmproxy/proxy/layers/test_websocket.py:101]: >  How about a `text` property that either works, or raises when called on a Binary message?

That's what I meant above with

> If `not is_text` the `text` property should crash or something tho?

What I don't like about this is that this means an imaginary user's addon works for five weeks and then it crashes because it sees a binary WebSocket for the first time. But on the other hand that's actually great, because if you don't check `is_text` then that's your fault and with the current design it would already result in unexpected behavior.

So yes, I think a `text` property that transparently handles the `encode`/`decode` and crashes when `not is_text` (both when reading and writing) sounds good to me.

Is it currently possible to change `is_text` (or the opcode rather, if this property is removed again)? I think it should be and the user instead should drop the message and inject a new one. Because convert from one to the other is not possible in all cases.

File: mitmproxy/websocket.py (hunk 5 above)
Comment (Prinzhorn) [reply mitmproxy/websocket.py:87]: > Going back to the drawing board, what would you expect `kill` to do? Promptly abort the connection, or just suppress the current message? I just now realize that we also don't have a good story for interception either.

First of all we might want to call it `drop` and not introduce more violence :wink: 

Calling `messages[-1].drop()` should not forward the current message to the other end but keep it in the `messages` list with `dropped == True`.

This also touches another area that I always found awkward: when handling the `websocket_message` event I care about the message that this event is about. But it doesn't have a special place and I have to access the messages list. This feels weird to me and even though there probably can't be any race conditions with the current design (?) I always wonder if `messages[-1]` is actually guaranteed to be the message that triggered `websocket_message`.

Maybe we can pass the message as an additional parameter to the event? That would make for a much nicer API because that's what the user cares about.

Comment (mhils) [reply mitmproxy/websocket.py:87]: `drop` is an excellent suggestion. 👍 

Do we need separate `.drop()` and `.dropped`? Or does a single `.drop` property do the job?



I agree on your point regarding `messages[-1]`, even though there are no race conditions at the moment. :) I think adding an extra argument is good, that would also allow us to make `websocket_message` non-blocking, i.e. we can continue to handle messages while one is intercepted. But that is future work, let's get the basics down here.

Comment (Prinzhorn) [reply mitmproxy/websocket.py:87]: > Do we need separate `.drop()` and `.dropped`? Or does a single `.drop` property do the job?

I thought it should be consistent with `flow.kill()` but can you even detect if a flow has been killed currently? I just peeked into the flow.py and there is no flag?

To me `recent_message.drop = True` feels less like it actually does something than doing `recent_message.drop()`, which is an imperative. But I guess we already use that pattern with `flow.stream = True`. So maybe we should turn this around and update `flow.drop = True` in a future change and remove `flow.kill()`?

Here's one counter argument: if there is just a single `message.drop` property you don't know if it has already been dropped or not, it depends on the context. Inside of `websocket_message` it means that it will be dropped after the event (or all events) has been processed, but if you inspect any older message in the list then `drop == True` means it had been dropped. The benefit of a property (see streaming) is that multiple chained events can work on it.

So maybe just a `drop` property and an additional new `active` (or `live`?) so that you know exactly what's up.

Comment (Prinzhorn) [reply mitmproxy/websocket.py:87]: > But that is future work, let's get the basics down here.

Future work definitely references #3396 :wink:

**Comment by mhils** [comment]
Thanks you two for the excellent feedback, that was really helpful. I propose we move the two outstanding discussion points -- drop/kill/... and `.text` -- to new issues. Any objections? This is good to go for my end. 😃

File: mitmproxy/proxy/layers/http/__init__.py (hunk 6 above)
Comment (Kriechi) [reply mitmproxy/proxy/layers/http/__init__.py:318]: https://github.com/python-hyper/wsproto/blob/55bf7a4ee1e85e3ebc93f04708327b0d0f193925/src/wsproto/handshake.py#L36

Comment (mhils) [reply mitmproxy/proxy/layers/http/__init__.py:318]: Yes that's good. We leave it hardcoded in the tests, so we have an obvious breakage if one of them ever changes.

File: mitmproxy/websocket.py [hunk 8]
```diff
@@ -1,168 +1,126 @@
...
+    """
+    A data container for everything related to a single WebSocket connection.
+    This is typically accessed as `mitmproxy.http.HTTPFlow.websocket`.
+    """
 
+    messages: List[WebSocketMessage]
+    """All `WebSocketMessage`s transferred over this connection."""
 
-class WebSocketFlow(flow.Flow):
+    close_by_client: Optional[bool] = None
```
Comment (Prinzhorn) [inline mitmproxy/websocket.py:98]: I know the uniform `close_` prefix is nice but `the close was done by the client` reads odd. How about `closed_by_client` which is short for `the WebSocket was closed by the client`

**Comment by Prinzhorn** [comment]
The WebSocket view keeps jumping to the top every time a new message arrives. This makes it basically impossible to work with while the connection is open. I hold down arrow -> it scrolls a bit -> message arrives -> I'm back at the top

File: mitmproxy/proxy/layers/http/__init__.py [hunk 9]
```diff
@@ -308,6 +309,25 @@ def send_response(self, already_streamed: bool = False):
...
+            self.context.options.websocket
+        )
+        if is_websocket:
+            # We need to set this before calling the response hook
+            # so that addons can determine if a WebSocket connection is following up.
+            self.flow.websocket = WebSocketData()
+            if self.flow.request.scheme == "http":
+                self.flow.request.scheme = "ws"
+            elif self.flow.request.scheme == "https":
+                self.flow.request.scheme = "wss"
```
Comment (Prinzhorn) [inline mitmproxy/proxy/layers/http/__init__.py:329]: I have hard feelings about this. `flow` is an HTTPFlow that happens to have a WebSocket using the same connection. If this is needed for the UI then the UI layer needs to handle this logic. But the scheme should never change.

```py
def requestheaders(flow):
    if(flow.request.pretty_host == 'echo.websocket.org'):
      print('requestheaders')
      print(flow.request.scheme)

def request(flow):
    if(flow.request.pretty_host == 'echo.websocket.org'):
      print('request')
      print(flow.request.scheme)

def responseheaders(flow):
    if(flow.request.pretty_host == 'echo.websocket.org'):
      print('responseheaders')
      print(flow.request.scheme)

def response(flow):
    if(flow.request.pretty_host == 'echo.websocket.org'):
      print('response')
      print(flow.request.scheme)
```

```
requestheaders
https

request
https

responseheaders
https

response
wss
```

giphy

Comment (Prinzhorn) [reply mitmproxy/proxy/layers/http/__init__.py:329]: Maybe a `flow.websocket.scheme` property?

File: mitmproxy/websocket.py [hunk 10]
```diff
@@ -1,168 +1,126 @@
...
-        It will not be sent to the other endpoint. This has no effect in streaming mode.
-        """
-        warnings.warn(
-            "WebSocketMessage.kill is deprecated, set an empty content instead.",
-            DeprecationWarning,
-            stacklevel=2,
-        )
-        # empty str or empty bytes.
-        self.content = type(self.content)()
+class WebSocketData(stateobject.StateObject):
```
Comment (Prinzhorn) [inline mitmproxy/websocket.py:89]: What do you think about `WebSocketDetails`? `data` has a certain connotation

**Comment by Prinzhorn** [comment]
Despite my comments this looks amazing :tada:
I'm done for now

File: mitmproxy/websocket.py (hunk 10 above)
Comment (mhils) [reply mitmproxy/websocket.py:89]: I picked "Data" over "Details" as I felt it's really not just the details, it's also the main part. 😛 
What connotation do you have for `data`?

Other alternatives:

 - WebSocketConnection: Bad idea IMO, that clashes with `mitmproxy.connection`.
 - WebSocketFlow: Bad idea IMO, this is not a top-level flow.
 - WebSocketPart
 - WebSocketExchange
 - WebSocketInfo
 - WebSocketMessages

Nothing really convinces me, I feel that "data" as in dataclass is the most straightforward.

Comment (Prinzhorn) [reply mitmproxy/websocket.py:89]: `WebSocketData` sounds great :+1: :smile:

**Comment by mhils** [comment]
> The WebSocket view keeps jumping to the top every time a new message arrives. This makes it basically impossible to work with while the connection is open. I hold down arrow -> it scrolls a bit -> message arrives -> I'm back at the top

Why do attempt actual real-world usage instead of synthetic tests with only one or two messages? Stop being weird. 🤷 

Jokes aside, yes this is bad. 😄 Agree that we need to fix it (likely also for TCP), but not as part of this PR.

**Comment by Prinzhorn** [comment]
> Why do attempt actual real-world usage instead of synthetic tests with only one or two messages?

We could also put "UI with support for up to five WebSocket messages" on the box and make it sound like that's a good thing :smile:

File: mitmproxy/proxy/layers/http/__init__.py (hunk 9 above)
Comment (mhils) [reply mitmproxy/proxy/layers/http/__init__.py]: I don't think it's that bad, but I don't feel as strongly. I've changed it to the UI layer. 👍 
(`flow.websocket.scheme` is not really necessary)
